src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/EntityComponent.scala

```diff
@@ -109,6 +117,90 @@ trait EntityComponent {
+          case "name" => ("e.name as sort_field, ", "", "", "")
+          case _ => (
+            """case
+              when sort_a.value_string is not null then sort_a.value_string
+              when sort_a.value_number is not null then sort_a.value_number
+              when sort_a.value_boolean is not null then sort_a.value_boolean
+              when e_ref.name is not null then e_ref.name
+              else null
+              end as sort_field, """,
+            """left outer join ENTITY_ATTRIBUTE sort_a on sort_a.owner_id = e.id
+              left outer join ENTITY sort_e_ref on sort_a.value_entity_ref = sort_e_ref.id """,
+            s"and sort_a.name = '$sortFieldName'",
+            ", sort_field"
+            )
+        }
+
+        sqlu"""create TEMPORARY table ENTITY_PAGINATION as
+        select e.id,
+        e.name,
+        #$sortColumn
+        group_concat(concat(ifnull(a.value_string, ''), ifnull(a.value_number, ''), ifnull(a.value_boolean, ''), ifnull(e_ref.name, '')) separator ' ') as filter_text
+
+        from ENTITY e
+        left outer join ENTITY_ATTRIBUTE a on a.owner_id = e.id
+        left outer join ENTITY e_ref on a.value_entity_ref = e_ref.id
+        #$sortJoin
+
+        where
+          e.entity_type = $entityType and
+          e.workspace_id = $workspaceId
+          #$sortConstraint
+
+        group by e.id, e.name #$sortOrder"""
+      }
+
+      def actionForPagination(workspaceContext: SlickWorkspaceContext, entityType: String, entityQuery: model.EntityQuery) = {
+        val baseSql = s"""select e.id, e.name, e.entity_type, e.workspace_id, e.record_version,
+          a.id, a.name, a.value_string, a.value_number, a.value_boolean, a.value_entity_ref, a.list_index,
+          e_ref.id, e_ref.name, e_ref.entity_type, e_ref.workspace_id, e_ref.record_version, p.name
+          from ENTITY e
+          left outer join ENTITY_ATTRIBUTE a on a.owner_id = e.id
+          left outer join ENTITY e_ref on a.value_entity_ref = e_ref.id"""
+
+        val filtersOption = entityQuery.filterTerms.map { _.split(" ").toSeq.map { term =>
+          sql"concat(pagination.name, ' ', pagination.filter_text) like ${'%' + term + '%'}"
+        }}
+
+        val filterSql = filtersOption match {
+          case None => sql""
+          case Some(filters) =>
+            concatSqlActions(sql"where ", reduceSqlActionsWithDelim(filters, sql" and "))
+        }
+
+        def order(alias: String) = sql" order by #$alias.sort_field #${SortDirections.toSql(entityQuery.sortDirection)} "
+
+        val paginationJoin = concatSqlActions(
+          sql""" join (select pagination.id, pagination.name, pagination.sort_field from ENTITY_PAGINATION pagination """,
+          filterSql,
+          order("pagination"),
+          sql" limit #${entityQuery.pageSize} offset #${(entityQuery.page-1) * entityQuery.pageSize} ) p on p.id = e.id "
+        )
+
```

Comment:
the following for loop outlines the general approach:
* populate a temp table with all the entities, all the attributes of an entity crammed into 1 column for filtering, and a column with the thing that should be sorted
* do the normal enitity query but join that to a sub select on the temp table with the appropriate filter, sort, limit and offset